libyul/optimiser/Solver.h

```diff
@@ -0,0 +1,92 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#pragma once
+
+#include <libyul/AST.h>
+#include <libyul/Dialect.h>
+#include <libyul/Exceptions.h>
+
+// because of instruction
+#include <libyul/backends/evm/EVMDialect.h>
+
+#include <libsmtutil/SolverInterface.h>
+
+#include <memory>
+
+namespace solidity::smtutil
+{
+
+class SolverInterface;
+class Expression;
+struct Sort;
+
+}
+
+namespace solidity::yul
+{
+
+/**
+ * Base class for SMT based optimization steps.
+ *
+ * Works best when used with AST in SSA form.
+ */
+class Solver
+{
+protected:
+	Solver(
+		std::set<YulString> const& _ssaVariables,
+		Dialect const& _dialect
+	);
+
+	/// Helper function that encodes VariableDeclaration
+	void encodeVariableDeclaration(VariableDeclaration const& _varDecl);
+
+	/// The encoding for a builtin. The type of encoding determines what we are
+	/// solving for.
+	virtual smtutil::Expression encodeEVMBuiltin(
+		evmasm::Instruction _instruction,
+		std::vector<Expression> const& _arguments
+	) = 0;
+
+	smtutil::Expression encodeExpression(Expression const& _expression);
+
+	smtutil::Expression int2bv(smtutil::Expression _arg) const;
+	smtutil::Expression bv2int(smtutil::Expression _arg) const;
+
+	smtutil::Expression newVariable();
+	virtual smtutil::Expression newRestrictedVariable(bigint _maxValue = (bigint(1) << 256) - 1);
+	std::string uniqueName();
+
+	virtual std::shared_ptr<smtutil::Sort> defaultSort() const;
+	virtual smtutil::Expression booleanValue(smtutil::Expression _value) const;
+	virtual smtutil::Expression constantValue(bigint _value) const;
+	virtual smtutil::Expression literalValue(Literal const& _literal) const;
+	virtual smtutil::Expression twosComplementToSigned(smtutil::Expression _value);
+	virtual smtutil::Expression signedToTwosComplement(smtutil::Expression _value);
```

Comment:
The two above can be `const` too

Reply:
Interesting, why aren't they static? Does it make sense to override them?

Reply:
Yeah, these functions can be both static and non-virtual. What do you think @leonardoalt

Reply:
Yep makes sense